### DISCUSSION TITLE
Hidration problem

### DISCUSSION THREAD

**Discussion by allestaire** [vercel/next.js#49513] [Help]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Damn dont know what just happen here, but this is a pain.
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

App directory (appDir: true)

### Link to the code that reproduces this issue

.

### To Reproduce

```js
// src/app/api/users/route.ts

import { NextResponse } from 'next/server'
import type { NextRequest } from 'next/server'

import { Client } from '../../../utils/prisma'

export async function GET(request: NextRequest) {
  // const json = await request.json()
  // console.log(request.url)
  const allUsers = await Client.user.findMany()
  // console.log(req)
  // return res.status(200).json({ name: 'John Doe' });
  return NextResponse.json(allUsers)
}

// src/app/dashboard/pages.jsx

'use client'
import React, { useEffect } from 'react'
import useSWR from 'swr'

...

const Dashboard = async () => {
  // const users = await getUsers()
  let rows = []
  const { data, error } = useSWR('/api/posts', fetch)
  console.log(data)

  }
  return (
   // trying to test fetching the data only
  )
}

export default Dashboard
```

### Describe the Bug

. When an api route exists hydration error occured

### Expected Behavior

.

**Comment by allestaire** [comment]
I just realy dont know if this is worth it,
I got so many issues when started exploring this framework.

Too many errors that did not occured on plain React
<img width="437" alt="Screen Shot 2023-05-08 at 9 06 50 PM" src="https://user-images.githubusercontent.com/68372409/236832364-55e7cbfe-935c-47ac-b78e-cebcf85657bc.png">

**Comment by allestaire** [comment]
This is not good, totally. tsk tsk

Im having a conclusion that client and server will not work on this

Im using prisma, this error occured when I start having `app/api/users`

**Comment by JesseKoldewijn** [comment]
Hey man!
To start of, sorry to hear you're encountering this issue. But you didn't really follow the issues template.
1. No `next info` output
2. No reproduction repo / codesandbox or something like it
3. No code snippets showing what your code looks like

Lacking these points make trouble shooting kinda of impossible so it would be greatly appreciated if you can share your output of `next info` and maybe a reproduction repo or codesandbox.

**Comment by allestaire** [comment]
Thanks for your response @JesseKoldewijn ,
And for being calm despite on my behaviour regarding on my issue

**Comment by JesseKoldewijn** [comment]
> Thanks for your response @JesseKoldewijn , And for being calm despite on my behaviour regarding on my issue

Do you have a reproduction repo or codesandbox for me so I could look into it for ya? 😄

**Comment by JesseKoldewijn** [comment]
> I just realy dont know if this is worth it, I got so many issues when started exploring this framework.
> 
> Too many errors that did not occured on plain React <img alt="Screen Shot 2023-05-08 at 9 06 50 PM" width="437" src="https://user-images.githubusercontent.com/68372409/236832364-55e7cbfe-935c-47ac-b78e-cebcf85657bc.png">

This could btw happen when you move from SPA based react to the new react + next.js standards (which is MPA + server-first)

**Reply by allestaire** [reply]
Hi @JesseKoldewijn this is the information on next.js on my local

```

    Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 21.3.0: Wed Jan  5 21:37:58 PST 2022; root:xnu-8019.80.24~20/RELEASE_ARM64_T8101
    Binaries:
      Node: 16.20.0
      npm: 8.19.4
      Yarn: 1.22.19
      pnpm: N/A
    Relevant packages:
      next: 13.4.0
      eslint-config-next: 13.4.0
      react: 18.2.0
      react-dom: 18.2.0

 warn  - Latest canary version not detected, detected: "13.4.0", newest: "13.4.2-canary.4".
        Please try the latest canary version (`npm install next@canary`) to confirm the issue still exists before creating a new issue.
        Read more - https://nextjs.org/docs/messages/opening-an-issue
```

**Reply by allestaire** [reply]
My question is, so meaning the framework itself is not configure by default to work both?
Or there are other things to do when working with api on next.js?

To tell, the API works
But when the frontend is access it causes hrydration problem making me confuse on why is that? I thought it could define or at least separate it when api files is defined.

**Reply by allestaire** [reply]
Note, this hydration problem occurs when api is visible, already confirmed when I removed the api folder and the hydration gone.

**Reply by JesseKoldewijn** [reply]
Edit: can you share a reproduction repo so I can look into it?

**Reply by JesseKoldewijn** [reply]
> My question is, so meaning the framework itself is not configure by default to work both? Or there are other things to do when working with api on next.js?
> 
> To tell, the API works But when the frontend is access it causes hrydration problem making me confuse on why is that? I thought it could define or at least separate it when api files is defined.

And I'm unsure what you mean by this. What do you mean with "configure by default to work both" and "when the frontend is access"?

**Reply by allestaire** [reply]
Will try to share a reproduction @JesseKoldewijn :)

**Comment by icyJoseph** [comment]
Please post your file tree structure, what URL are you trying to access, and some pseudo-coded indicating what is it your code is trying to do.

**Reply by JesseKoldewijn** [reply]
A reproduction repo might be the best solution to quickly find the culprit 😉

**Reply by allestaire** [reply]
Updated the ticket with **To Reproduce** content @JesseKoldewijn hope it could help 😄

**Reply by icyJoseph** [reply]
It was the async usage in a client component. Answered at the top.

**Reply by allestaire** [reply]
Hi @icyJoseph thanks for being here 😄 

I've got progress now, I notice that on the other page, the prisma works fine, no unlimited call, but on the other page the error occurs. Im still investigating the cause

**Reply by icyJoseph** [reply]
It was also pointed that you have `pages.jsx`, when it should be, `page.jsx`

**Comment by icyJoseph** [comment]
Client components can't be async. These are the regular components we used this far. The RSC spec has other ways to handle async data in client components.

**Reply by JesseKoldewijn** [reply]
this, and as far as I know it might also not handle the filename "pages" instead of "page" all to well (but dont quote me on that one)

**Reply by icyJoseph** [reply]
Ah yeah, that too. So many issues, hard to see through assumptions. Finally, some code brings clarity.

**Reply by JesseKoldewijn** [reply]
It also doesn't make it all to clear if the code in the dashboard page is so much all over the place that both TypeScript and eslint is trying to stab me when I was looking into it 😅

**Comment by JesseKoldewijn** [comment]
to be fair, I wasn't expecting it to be that obvious lol

**Comment by JesseKoldewijn** [comment]
Just to make it a bit easier on ya @allestaire maybe a good idea if you copy over this repo of mine to get a working setup.

https://github.com/JesseKoldewijn/quark.jkinsight.nl

Edit: Oops wrong repo

**Reply by JesseKoldewijn** [reply]
This repo got Next.js, App Router, Tailwind and TypeScript setup in there so it might help in your favor. (the next.js project is located in the client folder in the repo)

**Reply by allestaire** [reply]
Wow, thanks @JesseKoldewijn for all the help and also for @icyJoseph 

I got new update again, it seems like a component problem just like @icyJoseph point out,
I've done some testing on my local and it seems like the component that I have problem with is the only one causing the problem.
The thing is I have to point out which component on its sub component.

**Reply by JesseKoldewijn** [reply]
> Wow, thanks @JesseKoldewijn for all the help and also for @icyJoseph
> 
> I got new update again, it seems like a component problem just like @icyJoseph point out, I've done some testing on my local and it seems like the component that I have problem with is the only one causing the problem. The thing is I have to point out which component on its sub component.

As my and @icyJoseph mentioned your dashboard page wont work cuz of the following:

1. "use client" components can't be async
2. empty return on a component won't do you any favors
3. pages.jsx/tsx should be pages.jsx/tsx

**Reply by JesseKoldewijn** [reply]
> > Wow, thanks @JesseKoldewijn for all the help and also for @icyJoseph
> > I got new update again, it seems like a component problem just like @icyJoseph point out, I've done some testing on my local and it seems like the component that I have problem with is the only one causing the problem. The thing is I have to point out which component on its sub component.
> 
> As my and @icyJoseph mentioned your dashboard page wont work cuz of the following:
> 
> 1. "use client" components can't be async
> 2. empty return on a component won't do you any favors
> 3. pages.jsx/tsx should be pages.jsx/tsx

So @allestaire the dashboard component should look like this when renamed to page.jsx in your case:
```
'use client'
import React, { useEffect } from 'react'
import useSWR from 'swr'

const Dashboard = () => {
  let rows = []
  const { data, error } = useSWR('/api/posts', fetch)
  console.log(data)

  return <></>
}

export default Dashboard
```

**Reply by allestaire** [reply]
Im thinking this is the cause of nested pages? or nested element? 😓

**Reply by allestaire** [reply]
All of my component has 'use client'

**Reply by JesseKoldewijn** [reply]
> Im thinking this is the cause of nested pages? or nested element? 😓 

?

**Reply by allestaire** [reply]
Im building a route like this,
- dashboard (default route)
- dashboard/setting (other route)

But my implementation is creating `layout.jsx` then using the **props.children** to put the page component or so

**Reply by allestaire** [reply]
And I want that the dashboard layout should be used on all dashboard sub route without recreating or importing it again

**Reply by allestaire** [reply]
I guess I implemented it wrongly 😓

**Reply by JesseKoldewijn** [reply]
> All of my component has 'use client'

You're also using a async component on a "use client" component which wont work.

**Reply by JesseKoldewijn** [reply]
can you share the repo by chance? I can throw ya a PR with the fixes and comments where the issues are located 😉

**Reply by allestaire** [reply]
This is the reference I followed

https://nextjs.org/docs/app/building-your-application/routing/pages-and-layouts#nesting-layouts

**Reply by JesseKoldewijn** [reply]
> This is the reference I followed
> 
> https://nextjs.org/docs/app/building-your-application/routing/pages-and-layouts#nesting-layouts

In this case the layout is the least of your concerns at the moment.

**Reply by allestaire** [reply]
Ahhh I see @JesseKoldewijn  and @icyJoseph I see it now, hahahahaha
This line causes the hell out of me.
And I didnt remember that there is a **async** on my code

Thank you so much for helping me and pointing it out. Stupid me!!

<img width="500" alt="Screen Shot 2023-05-10 at 6 24 03 PM" src="https://github.com/vercel/next.js/assets/68372409/26823cbc-b994-49f5-9baa-60dd2d4db6d1">

**Reply by allestaire** [reply]
Thanks there are no more infinite call!!!

<img width="1417" alt="Screen Shot 2023-05-10 at 6 26 14 PM" src="https://github.com/vercel/next.js/assets/68372409/ed9124e4-58ce-412f-bb1b-9b8e6ec0e8f9">

**Reply by JesseKoldewijn** [reply]
Was that the only issue? Cuz tbh this must have been the most confusing issue I've ever commented on. Tip for in the future, please share a codesandbox or repo because this basically took 2 days longer than needed.

**Reply by allestaire** [reply]
Yes, Sorry @JesseKoldewijn , Im still new to this, and I guess my laziness to share the code cause this to took long.

**Reply by JesseKoldewijn** [reply]
> Yes, Sorry @JesseKoldewijn , Im still new to this, and I guess my laziness to share the code cause this to took long.

No problem, keep learning and make sure you're reading the correct docs (app router / pages) 👌

**Reply by allestaire** [reply]
Thanks @JesseKoldewijn this will set as experience and be remember
